crates/client/src/test_generator.rs

```diff
@@ -0,0 +1,109 @@
+use anyhow::Context;
+use fehler::{throw, throws};
+use std::path::{Path, PathBuf};
+use thiserror::Error;
+use tokio::fs;
+
+const TESTS_DIRECTORY: &str = "tests";
+const TESTS_FILE_NAME: &str = "test.rs";
+const CARGO_TOML: &str = "Cargo.toml";
+const ANCHOR_TOML: &str = "Anchor.toml";
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("invalid workspace")]
+    BadWorkspace,
+    #[error("must have current dir")]
+    MustHaveCurrentDir,
+    #[error("{0:?}")]
+    Io(#[from] std::io::Error),
+    #[error("{0:?}")]
+    Anyhow(#[from] anyhow::Error),
+}
+
+pub struct TestGenerator {
+    root: PathBuf,
+}
+
+impl TestGenerator {
+    pub fn new() -> Self {
+        Self {
+            root: std::env::current_dir().unwrap(),
+        }
+    }
+
+    /// Initializes the the `tests` directory with all the necessary files. Adds the `test.rs` file
+    /// and generates `Cargo.toml` with `dev-dependencies`
+    ///
+    /// # Errors
+    ///
+    /// It fails when:
+    /// - there is not a root directory (no `Anchor.toml` file)
+    #[throws]
+    pub async fn generate(&mut self) {
+        self.root = self.discover_root()?;
```

Comment:
Is it possible to initialize it in the `new` method? It would be nicer to know the root dir at the beginning. Do you have an idea how to do it correctly?

Reply:
I replaced the code with this which is maybe better: https://github.com/stefan-prokop-cz/trdelnik/commit/ceb57f4f10b2502bbd660c366c5f94788623665f

If you are ok with this suggestion, we can merge this PR and open the refactor PR or I can merge it into this one. It's up to you.

Reply:
I'm not sure what you mean.

Reply:
I just dont like the `mut` reference. We are trying to reassign the `self.root`'s value. I think the better solution is to run `self.discover_root()` inside the `new()` function and set the `self.root` value there. But it acts strange in my eyes. So thats why I introduced the mentioned commit to remove the `self.root` and just call `self.discover_root()` function and its result send as an argument.